### PR TITLE
feat: Add monitor keyword as a callback metric internally in model_checkpoint callback

### LINKED ISSUE #7260: automatic metric selection filename ModelCheckpoint
## 🚀 Feature

when configuring the `filename` for the `ModelCheckpoint`, it might be nice to have an option to automatically paste the metric used for monitoring.

### Motivation

This would simplify setting up the ModelCheckpoint saving by dynamically creating the filename template.

### Pitch

Reserve "{\_\_monitor\_\_}" to be automatically set to the variable used for monitoring by `monitor=`. For example:

```
    cp = ModelCheckpoint(
        dirpath=dirpath,
        filename='{epoch}-{step}-{__monitor__:.2f}',
        save_top_k=3,
        mode='max',
        monitor='val_accuracy',
        )
```

could save the checkpoint to `epoch=1-step=2-val_accuracy=0.99.ckpt`.

### PR DESCRIPTION
**Pull request by Lucklyric**
## What does this PR do?
Add monitor keyword as a callback metric internally in model_checkpoint callback. 

Fixes #7260 

## Before submitting
- [X] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Lucklyric** [comment]
When I check the tests in https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/checkpointing/test_model_checkpoint.py. There are many test cases set `monitor='early_stop_on'` which will throw `pytorch_lightning.utilities.exceptions.MisconfigurationException` when actually apply it in a real workflow.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -532,6 +532,11 @@ def format_checkpoint_name(self, metrics: Dict[str, _METRIC], ver: Optional[int]
             >>> os.path.basename(ckpt.format_checkpoint_name(dict(epoch=2, val_loss=0.123456)))
             'epoch=2-val_loss=0.12.ckpt'
             >>> ckpt = ModelCheckpoint(dirpath=tmpdir,
+            ... filename='{epoch}-{val_loss:.2f}-{monitor:.2f}',
+            ... monitor='val_loss')
+            >>> os.path.basename(ckpt.format_checkpoint_name(dict(epoch=2, val_loss=0.123456)))
+            'epoch=2-val_loss=0.12-monitor=0.12.ckpt'
```
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:538]: IMO, I would prefer filename='{epoch}-{val_loss:.2f}-{monitor:.2f}' to be rendered 'epoch=2-val_loss=0.12-val_loss=0.12.ckpt' where monitor is being converted to val_loss.

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:538]: So users could just do filename='{epoch}-{monitor:.2f}' and easily change monitor metric name without changing the filename.

Comment (Lucklyric) [reply pytorch_lightning/callbacks/model_checkpoint.py:538]: Hello, @tchaton. I was thinking about it as well, but would it result in inconsistent logic with other metrics? The name for all other metrics will simply be their template strings. I am not sure, but I am glad to update this PR accordingly if you think the auto conversion would be better.

Comment (Lucklyric) [reply pytorch_lightning/callbacks/model_checkpoint.py]: Hi @tchaton, I reconsidered the problem and now agree with you. Keeping the keyword 'monitor' leads to the possibility of confusion between experiments with different monitor metrics. I just updated the PR.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -504,6 +505,10 @@ def _format_checkpoint_name(
             for group in groups:
                 name = group[1:]
 
+                if (name == 'monitor' and monitor is not None):
+                    metrics['monitor'] = metrics[monitor]
+                    name = monitor
+
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:511]: @tchaton  @Lucklyric I also prefer @carmocca 's suggestion here: https://github.com/PyTorchLightning/pytorch-lightning/issues/7260#issuecomment-836619959

this keeps the callback simpler and avoids special casing like this. wdyt?

Comment (Lucklyric) [reply pytorch_lightning/callbacks/model_checkpoint.py:511]: I agree with it
